casper/src/main/scala/coop/rchain/casper/util/ProtoUtil.scala

```diff
@@ -36,13 +36,18 @@ object ProtoUtil {
     import coop.rchain.catscontrib.Catscontrib.ToBooleanF
     import cats.instances.option._
 
-    (candidateMetadata.blockHash == targetBlockHash).pure[F] ||^ {
-      for {
-        targetBlockOpt       <- dag.lookup(targetBlockHash)
-        mainParentOpt        = targetBlockOpt >>= (_.parents.headOption)
-        inMainParentChainOpt <- mainParentOpt.traverse(isInMainChain(dag, candidateMetadata, _))
-      } yield inMainParentChainOpt.getOrElse(false)
-    }
+    (candidateMetadata.blockHash == targetBlockHash).pure[F] ||^
+      dag.lookup(targetBlockHash).flatMap {
+        case Some(targetBlock) =>
+          if (targetBlock.blockNum <= candidateMetadata.blockNum) {
+            false.pure[F]
+          } else {
+            val mainParentOpt = targetBlock.parents.headOption
+            mainParentOpt.traverse(isInMainChain(dag, candidateMetadata, _)).map(_.getOrElse(false))
+          }
+        case None =>
+          false.pure[F]
+      }
```

Comment:
```suggestion
    (candidateMetadata.blockNum <= targetBlock.blockNum).pure[F] &&^ (
      (candidateMetadata.blockHash == targetBlockHash).pure[F] ||^ {
        for {
          targetBlockOpt       <- dag.lookup(targetBlockHash)
          mainParentOpt        = targetBlockOpt >>= (_.parents.headOption)
          inMainParentChainOpt <- mainParentOpt.traverse(isInMainChain(dag, candidateMetadata, _))
        } yield inMainParentChainOpt.getOrElse(false)
      }
    )
```

Reply:
Alternatively, we could even make the check only once `candidateMetadata.blockNum <= targetBlock.blockNum`, and then delegate to a recursive helper that omits the check.

Reply:
We have to look up `targetBlock` first (line 53) so I can't move the check out of that for unfortunately. It is also impossible to do this check only once (`targetBlock` is dynamic here and changes the further we go) unless you mean we can deduce how far we should go from the difference between block numbers of target and candidate blocks. If so, I would argue that it brings unnecessary complexity into the method.

Reply:
You're right, my comment about doing the check once is wrong :)

Still we should be able to avoid pattern matching and manual control flow using combinators. We could for example:

```scala
        targetBlockOpt       <- dag.lookup(targetBlockHash)
          .map(_.filter(_.blockNum >= candidateMetadata.blockNum))
)
```

It's a matter of taste though :)